File: built-in plugins dsp/mb-dist-fuzz.cpp
/**
 * MB Fuzz
 * Category : effect
 * Type     : distortion
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Vintage fuzz pedal
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_DIST_FUZZ_H
#define MB_DIST_FUZZ_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbDistFuzz : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-dist-fuzz";
    static constexpr const char* PLUGIN_NAME    = "MB Fuzz";
    static constexpr const char* PLUGIN_TYPE    = "distortion";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float fuzz = 0.7f;  // range [0, 1]
    float tone = 0.4f;  // range [0, 1]
    float mix = 1f;  // range [0, 1]
    };

    MbDistFuzz() = default;
    ~MbDistFuzz() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.fuzz = std::clamp(params.fuzz, 0f, 1f);
        params.tone = std::clamp(params.tone, 0f, 1f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Fuzz
        return input;
    }
};

#endif // MB_DIST_FUZZ_H

File: built-in plugins dsp/mb-dist-fuzz.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbDistFuzzPlugin: PluginDefinition = {
  id: "mb-dist-fuzz",
  slug: "mb-dist-fuzz",
  name: "MB Fuzz",
  category: "effect",
  type: "distortion",
  version: "1.0.0",
  description: "Vintage fuzz pedal",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "fuzz",
      name: "Fuzz",
      type: "float",
      defaultValue: 0.7,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "tone",
      name: "Tone",
      type: "float",
      defaultValue: 0.4,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "mix",
      name: "Mix",
      type: "float",
      defaultValue: 1,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { fuzz: 0.7, tone: 0.4, mix: 1 },
};

export default MbDistFuzzPlugin;
